internal/graph/check.go

```diff
@@ -981,6 +988,466 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
 	}
 }
 
+// recursiveMatchUserUsersetCommonData groups common parameters needed
+// for recursiveMatchUserUserset for convenience purpose.
+type recursiveMatchUserUsersetCommonData struct {
+	typesys                     *typesystem.TypeSystem
+	ds                          storage.RelationshipTupleReader
+	allowedUserTypeRestrictions []*openfgav1.RelationReference
+	userToUsersetMapping        storage.SortedSet
+	concurrencyLimit            int
+
+	// The following member are atomic/sync in anticipation
+	// that the algorithm will parallelize the lookup.
+	dsCount        *atomic.Uint32
+	visitedUserset *sync.Map
+}
+
+// recursiveMatchUserUsersetFunc defines a function that recursively evaluates whether objects' matches userToUsersetMapping.
+type recursiveMatchUserUsersetFunc func(ctx context.Context,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetCommonData) (*ResolveCheckResponse, error)
+
+func parallelizeRecursiveMatchUserUserset(ctx context.Context,
+	usersetItems []string,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetCommonData,
+	recursiveFunc recursiveMatchUserUsersetFunc,
+) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "parallelizeRecursiveMatchUserUserset")
+	defer span.End()
+	checkOutcomeChan := make(chan checkOutcome, commonParameters.concurrencyLimit)
+
+	pool := concurrency.NewPool(ctx, commonParameters.concurrencyLimit)
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+
+	wg := &sync.WaitGroup{}
```

Comment:
Why the need for this?

Reply:
Right, we don't need this.  The wg is removed.

Reply:
This was so that we waited for the goroutine to exit before returning from this function. We could have flakey tests if we remove the waiting